build a C interface to taskchampion

djmitche opened the pull request:
I thought I'd get a start on this, and so far it's going OK.

Among other things, this would make it possible to use TaskChampion as the "backend" to TaskWarrior.  It's not clear that's the best plan yet, but even if we don't do so I think this will be a nice addition.

dbr commented on lib/src/replica.rs:
```diff
@@ -0,0 +1,136 @@
+/// undone.
+#[no_mangle]
+pub extern "C" fn tc_replica_undo<'a>(rep: *mut TCReplica) -> i32 {
+    wrap(rep, |rep| Ok(if rep.undo()? { 1 } else { 0 }), -1)
```
Could maybe return some kind of enum instead of numbers? So on C side it would be end up being roughly identical, but would allow doing `tc_replica_undo(...) == TC_UNDO_SUCCESS`

dbr replied:
..although it probably makes sense to experiment with error handling on `tc_replica_sync` as that has more interesting set of errors to deal with (mainly "errors with with error messages")

djmitche replied:
I like the enum idea.  I don't know how many functions will have the equivalent of `Result<bool>`, so I'll need to think about how to generalize that.  In any case, the error message for any of these is handled with `tc_replica_error`, so that should be OK.

dbr reviewed (approved):
This looks good :partying_face:

Aside from the one error handling note, the C API looks about like I'd expect such a thing to look. I kept thinking "oh that API looks a bit accident prone, what if.." before remembering "C" :roll_eyes:

I guess the main test of this would be in actually trying to integrate it with TW (obviously with how it works, but also things like how it integrates with build-systems)

djmitche commented:
Thanks for the check!

How does the `TCString` support strike you?  It's basically Cow but with CStr, CString, and String.  It _seems_ like something like this would already exist, but a bunch of searching didn't find anything.

dbr commented on lib/src/string.rs:
```diff
@@ -0,0 +1,202 @@
+/// after the TCString is freed.  It's typically easiest to ensure this by using a static string.
+///
+/// NOTE: this function does _not_ take responsibility for freeing the C string itself.
+/// The underlying string once the TCString has been freed.  Among other times, TCStrings are
```
"The underlying string once the TCString has been freed."  sentence seems incomplete

dbr commented on integration-tests/build.rs:
```diff
@@ -0,0 +1,36 @@
+
+    let files = &[
+        "src/bindings_tests/test.c",
+        // keep this list in sync with integration-tests/src/bindings_tests/mod.rs and
```
Could prevent the repetition by generating the snippets to `include!(..)` in `bindings_test/mod.rs` etc - maybe even glob the test files (although that's more debatable)

dbr reviewed (commented):
Not sure how tedious it to implement but would be better to include some kind of plain-text .sql file and generate the .sqlite3 file from this (or maybe even generate it from the main Rust lib?), instead of checking in the slightly-opaque binary `integration-tests/test-db/taskchampion.sqlite3`

djmitche commented:
> Not sure how tedious it to implement but would be better to include some kind of plain-text .sql file and generate the .sqlite3 file from this (or maybe even generate it from the main Rust lib?), instead of checking in the slightly-opaque binary `integration-tests/test-db/taskchampion.sqlite3`

Oops, that's not supposed to be checked in :)

dbr commented:
> How does the `TCString` support strike you? It's basically Cow but with CStr, CString, and String. It _seems_ like something like this would already exist, but a bunch of searching didn't find anything.

It looks perfectly reasonable on first glance, and I agree it definitely seems like something that probably already exists (but such crates can be really hard to search for). @thomcc has a lot more experience with Rust strings+FFI things and may know of something?

djmitche commented:
Thanks!  I got directed toward `ustr` but that doesn't look useful (and certainly not the "never frees strings" part!).

thomcc commented:
I don't know of a great utility crate for strings in FFI (I was pretty happy with the string support I wrote for https://crates.io/crates/ffi-support, but it doesn't support everything your TCString does). I think what you have here is probably not something you'll find off the shelf on crates.io.

I'm sick now, so can't give a thorough review, but did a quick skim, and I will note a lot of the code here is unsound, at least technically — Many safe functions accept a raw pointer, and dereference it (or Box::from_raw it). I'm not sure how much that matters to you, though, It's possibly fine given that this seems like an application not a library, and it wasn't clear if these cases are ever public.

P.S. Surprised to see two folks I (vaguely) know interacting :)

djmitche commented:
:waves: I suppose Rust is a bit of a small world :)

My understanding is that FFI necessarily involves unsafe Rust, but I'm not sure how what I've done is unsound.  Is it specifically functions that aren't `export "C" { .. }` like
```rust
impl<'a> TCString<'a> {
    pub(crate) fn from_arg(tcstring: *mut TCString<'a>) -> Self {
        debug_assert!(!tcstring.is_null());
        *(unsafe { Box::from_raw(tcstring) })
    }
...
```
, and would marking that function `unsafe` address the unsoundess?  No need to answer now if you're not feeling well -- I'm sure this PR will be open for a while yet :)

thomcc commented:
> would marking that function unsafe address the unsoundess

Yes. Soundness is more or less "you can cause undefined behavior with this safe function". The issue I'm describing is just that the function isn't marked as unsafe.

djmitche commented:
OK, thanks -- that's easy to fix!  I'm also planning to go back to every `unsafe { .. }` and add a comment as to why it's safe.

thomcc commented:
Right, the thing here is that the reason it might be safe isn't known in the function — it just gets an arbitrary pointer as an argument, but there's nothing that guarantees the pointer's validity (you can do `3usize as *mut Foo` in safe code, for example, or more realistically, `Vec::<Foo>::new().as_ptr()` — ignoring the UAF you have in this expression, this pointer is not valid for any bytes).

So generally, something that just wraps an unsafe function in a safe wrapper is wrong (unless it exists just to call something over the FFI), unless the unsafe function should have been safe in the first place (and the ones in the stdlib are pretty good in this respect)

djmitche commented:
> Relatedly, I think there was a TODO comment somewhere in this PR about at least validating the strings, right

Yep, I think the issue is that CStr (and CString, I think) do not validate their inputs are utf-8.  Anyway, it's easy to test and fix :)

djmitche commented:
OK, I addressed the utf-8 issues.  Anywhere C gets a string from something that's not known to be utf-8, it's checked and an error raised before Rust sees the data as a string.

I did a little extra work to make sure that's raised when the TCString is _consumed_ rather than when it's constructed, just to avoid having to do extra error checking on the C calls to construct strings (so all of the tc_string_... constructors _never_ return NULL).  On the Rust side, the Strings/&str are wrapped in Result, so it's impossible to forget to check for errors :)

I also addressed the unsoundness.  The utility functions (`TCString::{from_arg,from_arg_ref}`) are now unsafe.  And every unsafe block has a comment indicating why it's safe.  A lot of that is "C pinky-swears it to be the case" but I've checked that the C comments (which appear in `libtaskchampion.h`) describe the nature of those pinky-swears.

thomcc commented:
I'm not sure how much you should care, but the `foo_ref` functions (`task_ref`, `rep_ref`, ...) have the soundness holes I mentioned as well (and strictly speaking, you should probably mark the `extern "C" fn`s that take pointer args as unsafe).

That said, again, I don't know how much benefit there is from being dogmatic — it is nice to keep the safety reasoning entirely local for safe functions though, rather than based on the caller, but for this kind of code you probably have to audit the whole thing for safety anyway, and there's not that much so that might be fine.

djmitche commented:
Oh, indeed, I missed those -- I was focused on strings.

At the moment, marking the extern "C" functions unsafe would result in the entire body being regarded as unsafe, which I do not want.

dbr reviewed (commented):
(notes from quick ready through of changes the other day; don't think this was complete)

djmitche commented on integration-tests/build.rs:
```diff
@@ -0,0 +1,51 @@
+    libtaskchampion.push("target");
+    libtaskchampion.push(env::var("PROFILE").unwrap());
+    libtaskchampion.push("deps");
+    libtaskchampion.push("libtaskchampion.so");
```
Oh, thanks! Somehow I didn't see that var.

djmitche replied:
Oh, I spoke too soon:
```
  thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: NotPresent', integration-tests/build.rs:11:55
```
I recall figuring out that's only set during compilation, or something like that.

dbr replied:
Ah, yep :cry:  https://github.com/rust-lang/cargo/pull/7325#issuecomment-528365640 "Build scripts are intended to have a very scoped output (just in their OUT_DIR),"

It links to https://github.com/rust-lang/cargo/issues/5457 which links to another issue which links to cargo-xtask as a possible more robust workaround, but I'd suspect it's not worth the effort until the `target/$PROFILE/...` thing proves to be an actual problem

djmitche replied:
Hm, yeah, maybe `cargo xtask generate` or something like that would be a good way to generate the `taskchampion.h` header.  At the moment, that's in Git, but generated by `build.rs`, which is probably not a good combination.

djmitche commented on lib/src/uuid.rs:
```diff
@@ -0,0 +1,93 @@
+
+/// Parse the given string as a UUID.  Returns false on failure.
+#[no_mangle]
+pub extern "C" fn tc_uuid_from_str<'a>(s: *mut TCString, uuid_out: *mut TCUuid) -> bool {
```
I thought `extern "C"` implied unsafe, but maybe not..

djmitche replied:
Ah, I knew I'd seen that somewhere.  Yet, `unsafe extern "C"` behaves differently -- in particular, with that change the unit test that calls `tc_strings_free` requires `unsafe { .. }`.  Anyway, `unsafe extern "C"` is nice and explicit.

djmitche commented:
It's getting there!  I'm pretty happy with the utility traits.  I think they make it pretty hard to mess up the safety, or do things that are confusing from C.  Luckily in this particular application, aside from TCString, everything is 'static and owned by the Rust side, so it's pretty straightforward.

Bits still to do:
 * sync (requires a new type to contain sync config)
 * working set (requires a new wrapper for WorkingSet)
 * taskmap (new type)
 * annotations (new list type)
 * udas (new list type)
 * legacy udas (new list type)

all of that is formulaic, just a lot of copy/pasting :)

djmitche commented:
OK, I think this is in a good place to get reviewed.  The only issue I know of right now is that it doesn't handle Windows paths, and probably the C tests won't work on Windows, either.  But, we don't currently claim Windows compat so I think it's OK to merge wit that functionality missing and add it when it's required.

djmitche commented:
I had forgotten I promised to use codegen instead of build.rs to make the header file, so I fixed that :)

djmitche commented:
I'm looking at a way to make TCString PassByValue, but that can come in a subsequent PR.  It will save a lot of small allocations!

djmitche commented:
OK, updated to make TCString pass-by-value (which was a major change!)

djmitche commented:
Looks like I need to work on build-system compatibility on mac, but otherwise I'd love a review of this.

djmitche commented:
Looks like this is https://github.com/alexcrichton/cc-rs/issues/250 -- it just _happens_ to work on linux :(

djmitche commented:
```
  = note: taskchampion.lib(getrandom-719baf4ba8413e3e.getrandom.5orpj5wj-cgu.4.rcgu.o) : error LNK2019: unresolved external symbol __imp_BCryptGenRandom referenced in function _ZN9getrandom3imp15getrandom_inner17hb74b95fb2783e364E
  ```
it looks like, on windows, a dependency on bcrypt has to be included.  Maybe something similar to https://github.com/corrosion-rs/corrosion/pull/120, but in this case it's occuring on Rust 1.47.0.

djmitche commented:
Woo, all green (except lint, somehow)

dbr commented:
Had a look over the code and all looked good, although it's a lot of changes so I ended up skimming a lot!

Two things, one trivial and one maybe less so:

1. I tried running the tests with `cargo-nextest` out of curiosity and it fails to find the `libtaskchampion.so` which is interesting (which is potentially a bug with that project as it I don't think it should change how compilation works at all, but also quite likely the way we find the library here might also be a quirk of regaulr cargo-test rather than any "official" file layout). Not worth looking into for this PR
2. Running the test executable under valgrind's memcheck claims some memory is leaked. I guess this could potentially an artifact of how the tests are run rather than a problem with the C API itself

```
$ $ env LD_LIBRARY_PATH=target/debug/deps/ valgrind --leak-check=full target/debug/deps/bindings-9af6c48072e7dff1
[...]
==503== 
==503== HEAP SUMMARY:
==503==     in use at exit: 297 bytes in 5 blocks
==503==   total heap usage: 7,883 allocs, 7,878 frees, 1,133,607 bytes allocated
==503== 
==503== 193 (64 direct, 129 indirect) bytes in 1 blocks are definitely lost in loss record 5 of 5
==503==    at 0x483577F: malloc (vg_replace_malloc.c:299)
==503==    by 0x48EB45B: alloc::alloc::alloc (alloc.rs:86)
==503==    by 0x48EB4E6: alloc::alloc::Global::alloc_impl (alloc.rs:166)
==503==    by 0x48EB709: <alloc::alloc::Global as core::alloc::Allocator>::allocate (alloc.rs:226)
==503==    by 0x48EB3BC: alloc::alloc::exchange_malloc (alloc.rs:316)
==503==    by 0x48EDB54: new<taskchampion::replica::TCReplica> (boxed.rs:191)
==503==    by 0x48EDB54: taskchampion::traits::PassByPointer::return_ptr (traits.rs:128)
==503==    by 0x48D547A: tc_replica_new_in_memory (replica.rs:134)
==503==    by 0x12CEB4: test_replica_get_task_not_found (replica.c:300)
==503==    by 0x12DF9E: UnityDefaultTestRun (unity.c:1846)
==503==    by 0x12D0C6: replica_tests (replica.c:324)
==503==    by 0x127329: integration_tests::bindings_tests::replica_tests (mod.rs:21)
==503==    by 0x126E0D: bindings::replica_tests (bindings.rs:21)
==503== 
==503== LEAK SUMMARY:
==503==    definitely lost: 64 bytes in 1 blocks
==503==    indirectly lost: 129 bytes in 2 blocks
==503==      possibly lost: 0 bytes in 0 blocks
==503==    still reachable: 104 bytes in 2 blocks
==503==         suppressed: 0 bytes in 0 blocks
==503== Reachable blocks (those to which a pointer was found) are not shown.
==503== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==503== 
==503== For counts of detected and suppressed errors, rerun with: -v
==503== ERROR SUMMARY: 3 errors from 3 contexts (suppressed: 0 from 0)
```

djmitche commented:
Oh, cool!  I thought about trying to run things under valgrind but assumed it would be much more complicated than that.  It looks like that is, indeed, a bug in the test (it doesn't call tc_replica_free) rather than an issue in the API.

Your first point is a good one, too -- in general, I don't know how best to go about getting all this linking stuff lined up, and I haven't found any good packages to copy from.  So I'm hoping we can land it and find out by people trying to link to it :)

Figuring out how to run the tests under valgrind in CI would be amazing, too -- at least worth an issue.

djmitche commented:
oops, i meant to squash that :/

djmitche commented:
Filed #337 and #338 as followups.
